### PR TITLE
Disable account after failed login attempts

### PR DESCRIPTION
**Pull request by rullzer**
Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

### CODE REVIEW

File: lib/AppInfo/Application.php [hunk 1]
```diff
@@ -27,8 +27,12 @@
 use OCA\Password_Policy\Capabilities;
 use OCA\Password_Policy\ComplianceService;
 use OCA\Password_Policy\Generator;
+use OCA\Password_Policy\HistoryCompliance;
```
Comment (blizzz) [inline lib/AppInfo/Application.php:30]: why, was it missing?

File: lib/Validator/UserNameValidator.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+use OC\HintException;
+
+class UserNameValidator implements IValidator {
+
+	public function __construct() {
+	}
+
+	public function validate(string $password): void {
+		// TODO: Implement validate() method.
```
Comment (blizzz) [inline lib/Validator/UserNameValidator.php:35]: o.O

File: lib/FailedLoginCompliance.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+		IConfig $config,
+		IUserManager $userManager,
+		PasswordPolicyConfig $passwordPolicyConfig) {
+		$this->config = $config;
+		$this->userManager = $userManager;
+		$this->passwordPolicyConfig = $passwordPolicyConfig;
+	}
+
+	public function failedLogin(string $uid) {
+		$user = $this->userManager->get($uid);
```
Comment (blizzz) [inline lib/FailedLoginCompliance.php:52]: the preLoginName2UserNames-or-something-like-this-magic is missing here

File: lib/AppInfo/Application.php (hunk 1 above)
Comment (kesselb) [reply lib/AppInfo/Application.php:30]:
```suggestion
```

File: lib/FailedLoginCompliance.php (hunk 3 above)
Comment (rullzer) [reply lib/FailedLoginCompliance.php:52]: No it is not. That is already done before the event is emitted

File: lib/Validator/UserNameValidator.php (hunk 2 above)
Comment (rullzer) [reply lib/Validator/UserNameValidator.php]: ah mmm leftover from something else

File: lib/FailedLoginCompliance.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	public function sucessfullLogin(IUser $user) {
+		$this->setAttempts($user->getUID(), 0);
+	}
+
+	private function getAttempts(string $uid): int {
+		return (int)$this->config->getUserValue($uid, 'password_policy', 'failedLoginAttempts', 0);
+	}
+
+	private function setAttempts(string $uid, int $attempts) {
+		return $this->config->setUserValue($uid, 'password_policy', 'failedLoginAttempts', $attempts);
```
Comment (blizzz) [inline lib/FailedLoginCompliance.php:86]:
```suggestion
	private function setAttempts(string $uid, int $attempts): void {
		$this->config->setUserValue($uid, 'password_policy', 'failedLoginAttempts', $attempts);
```

It doesn't return anything

File: lib/FailedLoginCompliance.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	public function __construct(
+		IConfig $config,
+		IUserManager $userManager,
+		PasswordPolicyConfig $passwordPolicyConfig) {
+		$this->config = $config;
+		$this->userManager = $userManager;
+		$this->passwordPolicyConfig = $passwordPolicyConfig;
+	}
+
+	public function failedLogin(string $uid) {
```
Comment (ChristophWurst) [inline lib/FailedLoginCompliance.php:51]: nit: a naming like `handleFailedLogin` or `onFailedLogin` would hint that this method reacts on something

**Pull request merged**